components/raftstore/src/store/peer_storage.rs

```diff
@@ -306,11 +307,40 @@ impl EntryCache {
     }
 
     fn trace_cached_entries(&mut self, entries: CachedEntries) {
+        let dangle_size = {
+            let mut guard = entries.entries.lock().unwrap();
+
+            let last_idx = guard.0.last().map(|e| e.index).unwrap();
+            let cache_front = match self.cache.front().map(|e| e.index) {
+                Some(i) => i,
+                None => u64::MAX,
+            };
+
+            let dangle_range = if last_idx < cache_front {
+                // All entries are not in entry cache.
+                0..guard.0.len()
+            } else if let Ok(i) = guard.0.binary_search_by(|e| e.index.cmp(&cache_front)) {
+                // Some entries are in entry cache.
+                0..i
+            } else {
+                // All entries are in entry cache.
+                0..0
+            };
+
+            let mut size = 0;
+            for e in &guard.0[dangle_range] {
+                size += bytes_capacity(&e.data) + bytes_capacity(&e.context);
+            }
+            guard.1 = size;
+            size
+        };
+
         let old_capacity = self.trace.capacity();
         self.trace.push_back(entries);
         let new_capacity = self.trace.capacity();
         let diff = Self::get_trace_vec_mem_size_change(new_capacity, old_capacity);
-        self.flush_mem_size_change(diff);
+
+        self.flush_mem_size_change(diff + dangle_size as i64);
```

Comment:
Should we need to subtract `dangle_size` here? As I understand it is actually dropped.

Reply:
Entries won't be put into `EntryCache` when they are fetched from `Storage`. That's the reason for calling it `dangle`. So here we trace the size of dangle entries in `EntryCache`. And then if they are evicted later, we can substract the size.

Reply:
Why should the entry cache size account for the dangle size? Why not separately track the size of entries sent to apply, and treating apply entries as including the dangle size?

Reply:
@hicqu